### ISSUE TITLE
Race condition/crash in tailer when reading from stdin

### ISSUE THREAD

**Issue by jvgutierrez** [google/mtail#289]
We're seeing some crashes on mtail 3.0.0 rc19 (version shipped with Debian Buster) running mtail with the following parameters: `-progs /etc/mtail -logs /dev/stdin`.

The program causing the issue seems to be this one:
```
counter varnish_resourceloader_inm
counter varnish_resourceloader_resp by status, cache_control, x_cache

/\thttp_method PURGE\t/ {
    # noop
} else {
    # Match: /w/load.php, /w/load.php?query, /w/load.php/path
    # Do not match: /w/loadXphp,  /w/load.phpXX?query
    /^side c\turl \/w\/load\.php([\?\/].*)?\tcache_status (?P<x_cache>.*)\thttp_status (?P<status>[0-9]{3}).*\tcache_control (-|.*max-age=(?P<maxage>[0-9]+)).*\tinm (?P<inm>[[:print:]]+)\t.*$/ {
        $inm != "-" {
            varnish_resourceloader_inm++
        }
        len($maxage) > 0 {
            $maxage == "300" {
                varnish_resourceloader_resp[$status]["short"][$x_cache]++
            } else {
                varnish_resourceloader_resp[$status]["long"][$x_cache]++
            }
        } else {
            varnish_resourceloader_resp[$status]["no"][$x_cache]++
        }
    }
}
```

Our current workaround is downgrading mtail to version 3.0.0-rc5, that effectively solves the issue.

**Comment by jaqx0r** [comment]
Thanks for the report! What does the crash look like?

**Comment by jvgutierrez** [comment]
```
Jan 23 17:15:36 cp4032 varnishmtail[3368]: panic: runtime error: index out of range
Jan 23 17:15:36 cp4032 varnishmtail[3368]: goroutine 1 [running]:
Jan 23 17:15:36 cp4032 varnishmtail[3368]: bytes.(*Buffer).WriteByte(0xc00029c000, 0xc00068e161, 0x0, 0x0)
Jan 23 17:15:36 cp4032 varnishmtail[3368]:         /usr/lib/go-1.11/src/bytes/buffer.go:271 +0xa8
Jan 23 17:15:36 cp4032 varnishmtail[3368]: bytes.(*Buffer).WriteRune(0xc00029c000, 0x61, 0xc91, 0x61, 0x1)
Jan 23 17:15:36 cp4032 varnishmtail[3368]:         /usr/lib/go-1.11/src/bytes/buffer.go:281 +0x14c
Jan 23 17:15:36 cp4032 varnishmtail[3368]: github.com/google/mtail/internal/tailer.(*File).Read(0xc000292400, 0x7ffc4b30ff26, 0xa)
Jan 23 17:15:36 cp4032 varnishmtail[3368]:         github.com/google/mtail/internal/tailer/file.go:197 +0xf9
Jan 23 17:15:36 cp4032 varnishmtail[3368]: github.com/google/mtail/internal/tailer.(*Tailer).openLogPath(0xc000283f80, 0x7ffc4b30ff26, 0xa, 0x0, 0x0, 0x0)
Jan 23 17:15:36 cp4032 varnishmtail[3368]:         github.com/google/mtail/internal/tailer/tail.go:251 +0x406
Jan 23 17:15:36 cp4032 varnishmtail[3368]: github.com/google/mtail/internal/tailer.(*Tailer).TailPath(0xc000283f80, 0x7ffc4b30ff26, 0xa, 0xc000211c78, 0x1)
Jan 23 17:15:36 cp4032 varnishmtail[3368]:         github.com/google/mtail/internal/tailer/tail.go:178 +0xd5
Jan 23 17:15:36 cp4032 varnishmtail[3368]: github.com/google/mtail/internal/tailer.(*Tailer).TailPattern(0xc000283f80, 0x7ffc4b30ff26, 0xa, 0xc000211d30, 0x1)
Jan 23 17:15:36 cp4032 varnishmtail[3368]:         github.com/google/mtail/internal/tailer/tail.go:160 +0x227
Jan 23 17:15:36 cp4032 varnishmtail[3368]: github.com/google/mtail/internal/mtail.(*Server).StartTailing(0xc0001a0b60, 0x794b0a, 0xc0001a0b60)
Jan 23 17:15:36 cp4032 varnishmtail[3368]:         github.com/google/mtail/internal/mtail/mtail.go:72 +0x150
Jan 23 17:15:36 cp4032 varnishmtail[3368]: github.com/google/mtail/internal/mtail.(*Server).Run(0xc0001a0b60, 0x8ef420, 0xc0001b4090)
Jan 23 17:15:36 cp4032 varnishmtail[3368]:         github.com/google/mtail/internal/mtail/mtail.go:335 +0x41
Jan 23 17:15:36 cp4032 varnishmtail[3368]: main.main()
Jan 23 17:15:36 cp4032 varnishmtail[3368]:         github.com/google/mtail/cmd/mtail/main.go:153 +0x6d7
Jan 23 17:15:36 cp4032 varnishmtail[3368]: Log overrun
```

**Comment by shdubsh** [comment]
Hi jaqx0r,

I work with @jvgutierrez and was able to reproduce it locally.  I compiled tag v3.0.0-rc19 with golang 1.13.4 and enabled the race detector.  The trace is consistent with what I'm seeing in Debian Buster (3.0.0~rc19-2).

```
==================
WARNING: DATA RACE
Write at 0x00c0001c6980 by goroutine 20:
  bytes.(*Buffer).WriteByte()
      /usr/local/go/src/bytes/buffer.go:264 +0x42
  bytes.(*Buffer).WriteRune()
      /usr/local/go/src/bytes/buffer.go:279 +0x1fd
  github.com/google/mtail/internal/tailer.(*File).Read()
      /go/src/github.com/google/mtail/internal/tailer/file.go:197 +0x139
  github.com/google/mtail/internal/tailer.(*File).Follow()
      /go/src/github.com/google/mtail/internal/tailer/file.go:143 +0x1f5
  github.com/google/mtail/internal/tailer.doFollow()
      /go/src/github.com/google/mtail/internal/tailer/tail.go:203 +0x3c
  github.com/google/mtail/internal/tailer.(*Tailer).handleLogEvent()
      /go/src/github.com/google/mtail/internal/tailer/tail.go:198 +0x1a1
  github.com/google/mtail/internal/tailer.(*Tailer).run()
      /go/src/github.com/google/mtail/internal/tailer/tail.go:293 +0xd5

Previous write at 0x00c0001c6980 by main goroutine:
  github.com/google/mtail/internal/tailer.(*File).sendLine()
      /usr/local/go/src/bytes/buffer.go:100 +0x272
  github.com/google/mtail/internal/tailer.(*File).Read()
      /go/src/github.com/google/mtail/internal/tailer/file.go:199 +0x14f
  github.com/google/mtail/internal/tailer.(*Tailer).openLogPath()
      /go/src/github.com/google/mtail/internal/tailer/tail.go:251 +0x4aa
  github.com/google/mtail/internal/tailer.(*Tailer).TailPath()
      /go/src/github.com/google/mtail/internal/tailer/tail.go:178 +0x214
  github.com/google/mtail/internal/tailer.(*Tailer).TailPattern()
      /go/src/github.com/google/mtail/internal/tailer/tail.go:160 +0x1f9
  github.com/google/mtail/internal/mtail.(*Server).StartTailing()
      /go/src/github.com/google/mtail/internal/mtail/mtail.go:72 +0x157
  github.com/google/mtail/internal/mtail.(*Server).Run()
      /go/src/github.com/google/mtail/internal/mtail/mtail.go:335 +0x77
  main.main()
      /go/src/github.com/google/mtail/cmd/mtail/main.go:153 +0x100c

Goroutine 20 (running) created at:
  github.com/google/mtail/internal/tailer.New()
      /go/src/github.com/google/mtail/internal/tailer/tail.go:79 +0x308
  github.com/google/mtail/internal/mtail.(*Server).initTailer()
      /go/src/github.com/google/mtail/internal/mtail/mtail.go:136 +0xfe
  github.com/google/mtail/internal/mtail.New()
      /go/src/github.com/google/mtail/internal/mtail/mtail.go:200 +0x335
  main.main()
      /go/src/github.com/google/mtail/cmd/mtail/main.go:148 +0xf70
==================
==================
WARNING: DATA RACE
Read at 0x00c0001c6960 by goroutine 20:
  bytes.(*Buffer).WriteByte()
      /usr/local/go/src/bytes/buffer.go:107 +0x54
  bytes.(*Buffer).WriteRune()
      /usr/local/go/src/bytes/buffer.go:279 +0x1fd
  github.com/google/mtail/internal/tailer.(*File).Read()
      /go/src/github.com/google/mtail/internal/tailer/file.go:197 +0x139
  github.com/google/mtail/internal/tailer.(*File).Follow()
      /go/src/github.com/google/mtail/internal/tailer/file.go:143 +0x1f5
  github.com/google/mtail/internal/tailer.doFollow()
      /go/src/github.com/google/mtail/internal/tailer/tail.go:203 +0x3c
  github.com/google/mtail/internal/tailer.(*Tailer).handleLogEvent()
      /go/src/github.com/google/mtail/internal/tailer/tail.go:198 +0x1a1
  github.com/google/mtail/internal/tailer.(*Tailer).run()
      /go/src/github.com/google/mtail/internal/tailer/tail.go:293 +0xd5

Previous write at 0x00c0001c6960 by main goroutine:
  github.com/google/mtail/internal/tailer.(*File).sendLine()
      /usr/local/go/src/bytes/buffer.go:98 +0x23e
  github.com/google/mtail/internal/tailer.(*File).Read()
      /go/src/github.com/google/mtail/internal/tailer/file.go:199 +0x14f
  github.com/google/mtail/internal/tailer.(*Tailer).openLogPath()
      /go/src/github.com/google/mtail/internal/tailer/tail.go:251 +0x4aa
  github.com/google/mtail/internal/tailer.(*Tailer).TailPath()
      /go/src/github.com/google/mtail/internal/tailer/tail.go:178 +0x214
  github.com/google/mtail/internal/tailer.(*Tailer).TailPattern()
      /go/src/github.com/google/mtail/internal/tailer/tail.go:160 +0x1f9
  github.com/google/mtail/internal/mtail.(*Server).StartTailing()
      /go/src/github.com/google/mtail/internal/mtail/mtail.go:72 +0x157
  github.com/google/mtail/internal/mtail.(*Server).Run()
      /go/src/github.com/google/mtail/internal/mtail/mtail.go:335 +0x77
  main.main()
      /go/src/github.com/google/mtail/cmd/mtail/main.go:153 +0x100c

Goroutine 20 (running) created at:
  github.com/google/mtail/internal/tailer.New()
      /go/src/github.com/google/mtail/internal/tailer/tail.go:79 +0x308
  github.com/google/mtail/internal/mtail.(*Server).initTailer()
      /go/src/github.com/google/mtail/internal/mtail/mtail.go:136 +0xfe
  github.com/google/mtail/internal/mtail.New()
      /go/src/github.com/google/mtail/internal/mtail/mtail.go:200 +0x335
  main.main()
      /go/src/github.com/google/mtail/cmd/mtail/main.go:148 +0xf70
==================
==================
WARNING: DATA RACE
Write at 0x00c0002600a1 by goroutine 20:
  bytes.(*Buffer).WriteByte()
      /usr/local/go/src/bytes/buffer.go:269 +0xc8
  bytes.(*Buffer).WriteRune()
      /usr/local/go/src/bytes/buffer.go:279 +0x1fd
  github.com/google/mtail/internal/tailer.(*File).Read()
      /go/src/github.com/google/mtail/internal/tailer/file.go:197 +0x139
  github.com/google/mtail/internal/tailer.(*File).Follow()
      /go/src/github.com/google/mtail/internal/tailer/file.go:143 +0x1f5
  github.com/google/mtail/internal/tailer.doFollow()
      /go/src/github.com/google/mtail/internal/tailer/tail.go:203 +0x3c
  github.com/google/mtail/internal/tailer.(*Tailer).handleLogEvent()
      /go/src/github.com/google/mtail/internal/tailer/tail.go:198 +0x1a1
  github.com/google/mtail/internal/tailer.(*Tailer).run()
      /go/src/github.com/google/mtail/internal/tailer/tail.go:293 +0xd5

Previous read at 0x00c0002600a0 by main goroutine:
  runtime.slicebytetostring()
      /usr/local/go/src/runtime/string.go:75 +0x0
  github.com/google/mtail/internal/tailer.(*File).sendLine()
      /usr/local/go/src/bytes/buffer.go:65 +0x36e
  github.com/google/mtail/internal/tailer.(*File).Read()
      /go/src/github.com/google/mtail/internal/tailer/file.go:199 +0x14f
  github.com/google/mtail/internal/tailer.(*Tailer).openLogPath()
      /go/src/github.com/google/mtail/internal/tailer/tail.go:251 +0x4aa
  github.com/google/mtail/internal/tailer.(*Tailer).TailPath()
      /go/src/github.com/google/mtail/internal/tailer/tail.go:178 +0x214
  github.com/google/mtail/internal/tailer.(*Tailer).TailPattern()
      /go/src/github.com/google/mtail/internal/tailer/tail.go:160 +0x1f9
  github.com/google/mtail/internal/mtail.(*Server).StartTailing()
      /go/src/github.com/google/mtail/internal/mtail/mtail.go:72 +0x157
  github.com/google/mtail/internal/mtail.(*Server).Run()
      /go/src/github.com/google/mtail/internal/mtail/mtail.go:335 +0x77
  main.main()
      /go/src/github.com/google/mtail/cmd/mtail/main.go:153 +0x100c

Goroutine 20 (running) created at:
  github.com/google/mtail/internal/tailer.New()
      /go/src/github.com/google/mtail/internal/tailer/tail.go:79 +0x308
  github.com/google/mtail/internal/mtail.(*Server).initTailer()
      /go/src/github.com/google/mtail/internal/mtail/mtail.go:136 +0xfe
  github.com/google/mtail/internal/mtail.New()
      /go/src/github.com/google/mtail/internal/mtail/mtail.go:200 +0x335
  main.main()
      /go/src/github.com/google/mtail/cmd/mtail/main.go:148 +0xf70
==================
panic: runtime error: index out of range [163] with length 0

goroutine 30 [running]:
bytes.(*Buffer).WriteByte(0xc0001c6960, 0x49ad38, 0xc00033a002, 0x0)
	/usr/local/go/src/bytes/buffer.go:269 +0x169
bytes.(*Buffer).WriteRune(0xc0001c6960, 0x38, 0xffe, 0x38, 0x1)
	/usr/local/go/src/bytes/buffer.go:279 +0x1fe
github.com/google/mtail/internal/tailer.(*File).Read(0xc0001be880, 0x0, 0xc000062470)
	/go/src/github.com/google/mtail/internal/tailer/file.go:197 +0x13a
github.com/google/mtail/internal/tailer.(*File).Follow(0xc0001be880, 0x20, 0x7ffe10af49f3)
	/go/src/github.com/google/mtail/internal/tailer/file.go:143 +0x1f6
github.com/google/mtail/internal/tailer.doFollow(0xc0001be880)
	/go/src/github.com/google/mtail/internal/tailer/tail.go:203 +0x3d
github.com/google/mtail/internal/tailer.(*Tailer).handleLogEvent(0xc0001cc230, 0x7ffe10af49f3, 0xf)
	/go/src/github.com/google/mtail/internal/tailer/tail.go:198 +0x1a2
github.com/google/mtail/internal/tailer.(*Tailer).run(0xc0001cc230, 0xc000090a80)
	/go/src/github.com/google/mtail/internal/tailer/tail.go:293 +0xd6
created by github.com/google/mtail/internal/tailer.New
	/go/src/github.com/google/mtail/internal/tailer/tail.go:79 +0x309
```

**Comment by jaqx0r** [comment]
Excellent, thank you both!

**Comment by jaqx0r** [comment]
I missed where you said you see this in rc11.  Can you try rc33 or rc34 or HEAD?  I am certain this is fixed.

**Comment by shdubsh** [comment]
Hi jaqx0r,

Mtail rc34 and HEAD do not panic, but also do not appear to process any logs when configured to read from /dev/stdin.  Requesting metrics from the metrics endpoint confirms this as no metrics from the mtail programs are exported.  Instance metrics like go_memstats_last_gc_time are present, however.

I can replicate the same panic in rc33 (from github):
```
panic: runtime error: index out of range

goroutine 23 [running]:
bytes.(*Buffer).WriteByte(0xc000455500, 0xc000592009, 0x0, 0x0)
	/usr/local/go/src/bytes/buffer.go:272 +0xa6
bytes.(*Buffer).WriteRune(0xc000455500, 0x9, 0x9b6, 0x9, 0x1)
	/usr/local/go/src/bytes/buffer.go:282 +0x14d
github.com/google/mtail/internal/tailer.(*File).Read(0xc0000b3980, 0xacc8a0, 0xc0001f0390, 0x0, 0x0)
	/go/src/github.com/google/mtail/internal/tailer/file.go:208 +0x1a9
github.com/google/mtail/internal/tailer.(*File).Follow(0xc0000b3980, 0xacc8a0, 0xc0001f0270, 0x0, 0x0)
	/go/src/github.com/google/mtail/internal/tailer/file.go:148 +0x297
github.com/google/mtail/internal/tailer.doFollow(0xacc8a0, 0xc0001f0120, 0xc0000b3980)
	/go/src/github.com/google/mtail/internal/tailer/tail.go:247 +0x43
github.com/google/mtail/internal/tailer.(*Tailer).ProcessFileEvent(0xc000127380, 0xacc8a0, 0xc0001f0120, 0x2, 0x7ffd8d0a09f8, 0xa)
	/go/src/github.com/google/mtail/internal/tailer/tail.go:242 +0x201
github.com/google/mtail/internal/watcher.(*LogWatcher).sendWatchedEvent(0xc0000b29c0, 0xc0004554d0, 0x2, 0x7ffd8d0a09f8, 0xa)
	/go/src/github.com/google/mtail/internal/watcher/log_watcher.go:97 +0x9c
github.com/google/mtail/internal/watcher.(*LogWatcher).sendEvent(0xc0000b29c0, 0x2, 0x7ffd8d0a09f8, 0xa)
	/go/src/github.com/google/mtail/internal/watcher/log_watcher.go:91 +0xef
github.com/google/mtail/internal/watcher.(*LogWatcher).runEvents(0xc0000b29c0)
	/go/src/github.com/google/mtail/internal/watcher/log_watcher.go:208 +0x1d2
created by github.com/google/mtail/internal/watcher.NewLogWatcher
	/go/src/github.com/google/mtail/internal/watcher/log_watcher.go:72 +0x10b
```
For reference, I am testing it like this:
```
$ cat varnishnmca.sample.output | ./mtail_v3.0.0-rc33_linux_amd64 -logtostderr -progs $(pwd)/conf -logs /dev/stdin
```
I observed that the first version to demonstrate this behavior was rc16.  Not sure if that helps at all though.

**Comment by jaqx0r** [comment]
The dev/stdin issue sounds like a different issues, can you file a bug? Or
remind me of an existing one...

**Comment by jaqx0r** [comment]
Thanks for isolating the version that stopped functioning as expected as
well.

**Comment by shdubsh** [comment]
Hi jaqx0r,

@jvgutierrez and I are colleagues working on the same problem.  We both experience the same panic when piping logs into mtail that are not backed by a regular file.

**Comment by jaqx0r** [comment]
Yeah I understood that; I misinterpreted the issue as being related to the
program compiler, not the tailer.

**Comment by shdubsh** [comment]
I had a chance to poke around at this a bit and confirmed the issue has been corrected in rc34 and onward.

Based on the behavior of rc34, I guess the tailer doesn't process logs until the watcher registers an event.  For example, `cat logfile | mtail -v 2 -logtostderr -progs $(pwd)/conf -logs /dev/stdin` doesn't work because the watcher doesn't pick up any writes, however `tail -f intermediate_log | mtail -logs /dev/stdin` coupled with a `cat logfile > intermediate_log` elsewhere triggers the watcher event and processes the logs appropriately.

Debian Buster (rc19) and Sid (rc24) are still affected, but the workaround I wrote does not apply to the current state of mtail.  Feel free to close at your discretion.
